rust/bin-shared/tests/network_notifiers.rs

```diff
@@ -0,0 +1,50 @@
+use firezone_bin_shared::{new_dns_notifier, new_network_notifier, platform::DnsControlMethod};
+use futures::{task::noop_waker, Future};
+use std::{
+    pin::pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tokio::time::timeout;
+
+/// Smoke test for the DNS and network change notifiers
+///
+/// Turn them on, wait a second, turn them off.
+/// This tests that the threads quit gracefully when we call `close`, and they don't crash on startup.
+#[tokio::test]
+async fn notifiers() {
+    firezone_logging::test_global("debug");
+    let tokio_handle = tokio::runtime::Handle::current();
+
+    let mut dns = new_dns_notifier(tokio_handle.clone(), DnsControlMethod::default())
+        .await
+        .unwrap();
+    let mut net = new_network_notifier(tokio_handle, DnsControlMethod::default())
+        .await
+        .unwrap();
+
+    tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+
+    // The notifiers always notify once they start listening for changes, to avoid gaps during startup.
+    timeout(Duration::from_secs(1), dns.notified())
+        .await
+        .unwrap()
+        .unwrap();
+    timeout(Duration::from_secs(1), net.notified())
+        .await
+        .unwrap()
+        .unwrap();
+
+    // After that first notification, we shouldn't get any other notifications during a normal unit test.
+
+    let waker = noop_waker();
+    let mut ctx = Context::from_waker(&waker);
+
+    assert!(matches!(pin!(dns.notified()).poll(&mut ctx), Poll::Pending));
+    assert!(matches!(pin!(net.notified()).poll(&mut ctx), Poll::Pending));
```

Comment:
You might find this one handy: https://docs.rs/futures/latest/futures/future/trait.FutureExt.html#method.now_or_never